copier/tools.py

```diff
@@ -172,3 +173,56 @@ def cleanup(self):
     @staticmethod
     def _robust_cleanup(name):
         shutil.rmtree(name, ignore_errors=False, onerror=handle_remove_readonly)
+
+
+def get_lines(content: bytes) -> StrSeq:
+    # Assume UTF-8 encoding for text files. Since Jinja2 is already doing this
+    # for templates, there  probably isn't much harm in also doing it for all
+    # text files. The worst case should be a failure to show the diffs. An
+    # alternative is to add chardet support.
+    return content.decode("utf-8").split(os.linesep)
+
+
+def is_binary(content: bytes) -> bool:
+    """Return a boolean indicating if the content is a binary file."""
+    return b"\x00" in content or b"\xff" in content
+
+
+def color_print(color: int, msg: str):
+    print(f"{color}{msg}{colorama.Fore.RESET}")
+
+
+def print_diff(dst_relpath: Path, original_content: bytes, new_content: bytes):
+    try:
+        original_lines = get_lines(original_content)
+    except Exception:
+        color_print(
+            colorama.Fore.YELLOW, "diff unavailable: unable to determine encoding"
+        )
+        return
+    try:
+        new_lines = get_lines(new_content)
+    except Exception:
+        color_print(
+            colorama.Fore.YELLOW, "diff unavailable: unable to determine encoding"
+        )
+        return
+
+    indent = " " * 6
+    for line in unified_diff(
+        original_lines,
+        new_lines,
+        f"old/{dst_relpath}",
+        f"new/{dst_relpath}",
+        lineterm="",
+    ):
+        if line.startswith("---") or line.startswith("+++"):
+            color_print(colorama.Fore.YELLOW, f"{indent}{line}")
+        elif line.startswith("@@"):
+            color_print(colorama.Fore.MAGENTA, f"{indent}{line}")
+        elif line.startswith("-"):
+            color_print(colorama.Fore.RED, f"{indent}{line}")
+        elif line.startswith("+"):
+            color_print(colorama.Fore.GREEN, f"{indent}{line}")
```

Comment:
https://python-prompt-toolkit.readthedocs.io/en/master/pages/printing_text.html#pygments-token-text-tuples with a diff lexer should save a lot of code around here.

Reply:
So I was going down this path first, but it seemed very difficult to get the result I wanted. I went ahead and finished that work a put together a [gist showing the differences](https://gist.github.com/dstanek/aad702dbc18ff54df6c41f54fe782f33). `orig.diff` is the code from this PR and `new.diff` uses `pygments` and `prompt_toolkit`.

One reason the new code is much larger is that I wanted to have the two-line file header yellow like it is with git. This meant that I had to extend the lexer. In order to add the extra indentation, I had to also extend the `PygmentsTokens` class to inject whitespace. There may be better ways to accomplish these tasks, but I'm never had to dip into the bowels of these libraries before.

Reply:
I'll update this implementation here since that's what I'm actively using and I think I can make it cleaner and more concise. If you like the other way better I can switch to that later.